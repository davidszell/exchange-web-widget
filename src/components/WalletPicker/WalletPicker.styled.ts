import styled from 'styled-components';

export const Container = styled.ul`
  list-style: none;
  padding: 0;
`;

export const WalletInfo = styled.li`
  display: flex;
  flex-direction: row;
  gap: 40px;
  border-bottom: 1px solid #ccc;
  cursor: pointer;
  color: #222;
  padding: 10px 40px;

  &:hover {
    color: #157ae2;
  }
`;

export const Name = styled.p`
  display: flex;
  align-items: center;
  margin: 0;
  font-size: 1.2em;
  font-weight: 500px;
`;

export const InfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 5px;
  flex: 1;
`;

export const LongName = styled.p`
  margin: 0;
`;

export const Balance = styled.p`
  margin: 0;
  font-size: 0.8em;
`;
