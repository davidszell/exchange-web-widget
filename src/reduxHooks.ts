import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { RootState } from "./store";

export const useAppDispatch = (): ThunkDispatch<RootState, any, AnyAction> => useDispatch<ThunkDispatch<RootState, any, AnyAction>>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector