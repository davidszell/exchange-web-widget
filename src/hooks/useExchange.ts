import { useEffect, useState } from 'react';
import { doExchange, fetchExchangeRates } from '../actions';
import { useAppDispatch, useAppSelector } from './reduxHooks';
import { ExchangeRatesType, WalletType } from '../types';
import { formatFloatRound } from '../helpers';

export type ExchangeActionType = 'buy' | 'sell';

const useExchange = () => {
  const wallets: WalletType[] = useAppSelector((state) => state.wallets);
  const exchangeRates: ExchangeRatesType = useAppSelector((state) => state.exchangeRates);

  const [sourceWallet, setSourceWallet] = useState<WalletType>(wallets[0]);
  const [destinationWallet, setDestinationWallet] = useState<WalletType>(wallets[1]);
  const [sourceAmount, setSourceAmount] = useState<number | null>(null);
  const [destinationAmount, setDestinationAmount] = useState<number | null>(null);
  const [exchangeAction, setExchangeAction] = useState<ExchangeActionType>('sell');
  const [isExchangeDisabled, setExchangeDisabled] = useState<boolean>(true);
  const [exchangeRate, setExchangeRate] = useState<number | null>(null);

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchExchangeRates());

    const interval = setInterval(() => {
      dispatch(fetchExchangeRates());
    }, 10000);

    return () => clearInterval(interval);
  }, [dispatch]);

  useEffect(() => {
    setSourceWallet((prevWallet) => wallets.find((wallet) => wallet.name === prevWallet.name) || wallets[0]);
    setDestinationWallet((prevWallet) => wallets.find((wallet) => wallet.name === prevWallet.name) || wallets[1]);
  }, [wallets]);

  useEffect(() => {
    setExchangeDisabled(
      sourceAmount === null
      || destinationAmount === null
      || (exchangeAction === 'sell' && sourceAmount > sourceWallet.balance)
      || (exchangeAction === 'buy' && destinationAmount > destinationWallet.balance),
    );
  }, [exchangeAction, sourceAmount, sourceWallet.balance, destinationAmount, destinationWallet.balance]);

  useEffect(() => {
    if (Object.keys(exchangeRates.rates).length === 0) {
      setExchangeRate(null);
      return;
    }

    const sourceRate = exchangeRates.rates[sourceWallet.name];
    const destinationRate = exchangeRates.rates[destinationWallet.name];

    if (sourceRate == null || destinationRate == null) {
      setExchangeRate(null);
      return;
    }

    if (sourceWallet.name === exchangeRates.base) {
      setExchangeRate(exchangeRates.rates[destinationWallet.name]);
    } else {
      setExchangeRate(destinationRate / sourceRate);
    }
  }, [sourceWallet, destinationWallet, exchangeRates]);

  const clearAmounts = () => {
    setSourceAmount(null);
    setDestinationAmount(null);
  };

  const handleSourceWalletChange = (walletName: string) => {
    if (sourceWallet.name === walletName) {
      return;
    }

    const newWallet = wallets.find((item) => item.name === walletName);

    if (!newWallet) {
      return;
    }

    if (destinationWallet.name === walletName) {
      setDestinationWallet(sourceWallet);
    }

    setSourceWallet(newWallet);
    clearAmounts();
  };

  const handleDestinationWalletChange = (walletName: string) => {
    if (destinationWallet.name === walletName) {
      return;
    }

    const newWallet = wallets.find((item) => item.name === walletName);

    if (!newWallet) {
      return;
    }

    if (sourceWallet.name === walletName) {
      setSourceWallet(destinationWallet);
    }

    setDestinationWallet(newWallet);
    clearAmounts();
  };

  const handleSourceAmountChange = (newAmount: number | null) => {
    if (exchangeRate !== null && newAmount !== null && newAmount > 0) {
      setSourceAmount(newAmount);
      setDestinationAmount(formatFloatRound(newAmount * exchangeRate));
    } else {
      clearAmounts();
    }
  };

  const handleDestinationAmountChange = (newAmount: number | null) => {
    if (exchangeRate !== null && newAmount !== null && newAmount > 0) {
      setDestinationAmount(newAmount);
      setSourceAmount(formatFloatRound(newAmount / exchangeRate));
    } else {
      clearAmounts();
    }
  };

  const toggleExchangeAction = () => {
    setExchangeAction((prevExchangeAction) => (prevExchangeAction === 'sell' ? 'buy' : 'sell'));
  };

  const handleExchange = () => {
    if (!isExchangeDisabled) {
      dispatch(doExchange(sourceWallet.name, destinationWallet.name, exchangeAction, sourceAmount, destinationAmount));
      clearAmounts();
    }
  };

  return {
    exchangeAction,
    exchangeRate,
    isExchangeDisabled,
    sourceWallet,
    destinationWallet,
    sourceAmount,
    destinationAmount,
    handleExchange,
    handleSourceWalletChange,
    handleDestinationWalletChange,
    handleSourceAmountChange,
    handleDestinationAmountChange,
    toggleExchangeAction,
  };
};

export default useExchange;
