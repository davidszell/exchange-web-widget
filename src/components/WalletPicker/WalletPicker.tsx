import React from 'react';
import { useAppSelector } from '../../hooks/reduxHooks';
import { WalletType } from '../../types';
import {
  Balance, Container, InfoContainer, LongName, Name, WalletInfo,
} from './WalletPicker.styled';

type WalletPickerProps = {
  handleWalletChange: (newWallet: string) => void
};

const WalletPicker = ({ handleWalletChange }: WalletPickerProps): JSX.Element => {
  const wallets: WalletType[] = useAppSelector((state) => state.wallets);

  return (
    <Container>
      {wallets.map((item) => (
        <WalletInfo key={item.name} onClick={() => handleWalletChange(item.name)}>
          <Name data-testid="name">{item.name}</Name>
          <InfoContainer>
            <LongName data-testid="longName">{item.longName}</LongName>
            <Balance data-testid="balance">
              {item.symbol}
              {' '}
              {item.balance}
            </Balance>
          </InfoContainer>
        </WalletInfo>
      ))}
    </Container>
  );
};

export default WalletPicker;
