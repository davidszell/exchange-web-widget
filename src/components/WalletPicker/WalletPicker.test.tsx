import { fireEvent, render, screen, within } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import WalletPicker from './WalletPicker';
import * as reduxHooks from '../../hooks/reduxHooks';

describe('WalletPicker', () => {
    const useSelectorMock = jest.spyOn(reduxHooks, 'useAppSelector')

    beforeEach(() => {
        useSelectorMock.mockImplementation(selector => selector(mockStore));
    })
    afterEach(() => {
        useSelectorMock.mockClear()
    })

    const mockStore = {
        wallets: [
            {
                name: "EUR",
                balance: 10,
                longName: "Euros",
                symbol: "€"
            },
            {
                name: "GBP",
                balance: 20,
                longName: "British Pounds",
                symbol: "£"
            },
            {
                name: "USD",
                balance: 30,
                longName: "US Dollars",
                symbol: "$"
            }
        ],
        exchangeRates: {}
    };

    it('matches snapshot', () => {
        const tree = renderer
            .create(<WalletPicker handleWalletChange={() => {}} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('renders a list', () => {
        render(<WalletPicker handleWalletChange={() => {}} />)
        const list = screen.getByRole('list')
        expect(list).toBeDefined()
    })

    it('lists all wallets from store', () => {
        render(<WalletPicker handleWalletChange={() => {}} />)
        const list = screen.getByRole('list')
        const items = within(list).getAllByRole('listitem')
        expect(items.length = 3)
    })

    it('renders wallet item name', async () => {
        render(<WalletPicker handleWalletChange={() => {}} />)
        const list = screen.getByRole('list')
        const items = within(list).getAllByRole('listitem')
        const name = within(items[0]).getByTestId('name')
        expect(name).toHaveTextContent(mockStore.wallets[0].name)
    })

    it('renders wallet item longname', () => {
        render(<WalletPicker handleWalletChange={() => {}} />)
        const list = screen.getByRole('list')
        const items = within(list).getAllByRole('listitem')
        const name = within(items[0]).getByTestId('longName')
        expect(name).toHaveTextContent(mockStore.wallets[0].longName)
    })

    it('renders wallet item balance', () => {
        const balanceText = `${mockStore.wallets[0].symbol} ${mockStore.wallets[0].balance}`
        render(<WalletPicker handleWalletChange={() => {}} />)
        const list = screen.getByRole('list')
        const items = within(list).getAllByRole('listitem')
        const name = within(items[0]).getByTestId('balance')
        expect(name).toHaveTextContent(balanceText)
    })

    it('calls handleWalletChange function when clicking on list item', () => {
        const handleWalletChangeMock = jest.fn()
        render(<WalletPicker handleWalletChange={handleWalletChangeMock} />)
        const list = screen.getByRole('list')
        const items = within(list).getAllByRole('listitem')
        fireEvent.click(items[0])
        expect(handleWalletChangeMock).toBeCalled()
    })

    it('calls handleWalletChange function with the wallet name', () => {
        const handleWalletChangeMock = jest.fn()
        render(<WalletPicker handleWalletChange={handleWalletChangeMock} />)
        const list = screen.getByRole('list')
        const items = within(list).getAllByRole('listitem')
        fireEvent.click(items[0])
        expect(handleWalletChangeMock).toBeCalledWith(mockStore.wallets[0].name)
    })
})