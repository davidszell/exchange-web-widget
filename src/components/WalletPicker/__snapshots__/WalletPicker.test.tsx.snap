// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletPicker matches snapshot 1`] = `
.c0 {
  list-style: none;
  padding: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 40px;
  border-bottom: 1px solid #ccc;
  cursor: pointer;
  color: #222;
  padding: 10px 40px;
}

.c1:hover {
  color: #157ae2;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0;
  font-size: 1.2em;
  font-weight: 500px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 5px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4 {
  margin: 0;
}

.c5 {
  margin: 0;
  font-size: 0.8em;
}

<ul
  className="c0"
>
  <li
    className="c1"
    onClick={[Function]}
  >
    <p
      className="c2"
      data-testid="name"
    >
      EUR
    </p>
    <div
      className="c3"
    >
      <p
        className="c4"
        data-testid="longName"
      >
        Euros
      </p>
      <p
        className="c5"
        data-testid="balance"
      >
        €
         
        10
      </p>
    </div>
  </li>
  <li
    className="c1"
    onClick={[Function]}
  >
    <p
      className="c2"
      data-testid="name"
    >
      GBP
    </p>
    <div
      className="c3"
    >
      <p
        className="c4"
        data-testid="longName"
      >
        British Pounds
      </p>
      <p
        className="c5"
        data-testid="balance"
      >
        £
         
        20
      </p>
    </div>
  </li>
  <li
    className="c1"
    onClick={[Function]}
  >
    <p
      className="c2"
      data-testid="name"
    >
      USD
    </p>
    <div
      className="c3"
    >
      <p
        className="c4"
        data-testid="longName"
      >
        US Dollars
      </p>
      <p
        className="c5"
        data-testid="balance"
      >
        $
         
        30
      </p>
    </div>
  </li>
</ul>
`;
